File: core/models/blockchain.py
from sqlalchemy import String, Boolean, DateTime, func, BigInteger, ForeignKey
from sqlalchemy.dialects.mysql import TEXT
from sqlalchemy.orm import mapped_column, relationship, Mapped

from core.db import Base
from core.dtos.base import BaseNftCollectionMetadataDTO, BaseNftItemMetadataDTO
from core.models.fields import (
    BlockchainAddressRawField,
    NftItemMetadataField,
    NftCollectionMetadataField,
)


class Jetton(Base):
    __tablename__ = "jetton"

    address = mapped_column(BlockchainAddressRawField, primary_key=True)
    name = mapped_column(String(255), nullable=False)
    description = mapped_column(TEXT, nullable=True)
    symbol = mapped_column(String(255), nullable=False)
    total_supply = mapped_column(BigInteger, nullable=False)
    logo_path = mapped_column(String(290), nullable=True)
    is_enabled = mapped_column(Boolean, nullable=False, default=True)
    created_at = mapped_column(DateTime(timezone=True), server_default=func.now())

    telegram_chat_jettons = relationship(
        "TelegramChatJetton",
        back_populates="jetton",
    )

    def __repr__(self):
        return f"<Jetton(address={self.address}, name={self.name})>"


class NFTCollection(Base):
    __tablename__ = "nft_collection"

    address = mapped_column(BlockchainAddressRawField, primary_key=True)
    name = mapped_column(String(255), nullable=False)
    description = mapped_column(TEXT, nullable=True)
    logo_path = mapped_column(String(290), nullable=True)
    is_enabled = mapped_column(Boolean, nullable=False, default=True)
    blockchain_metadata: Mapped[BaseNftCollectionMetadataDTO] = mapped_column(
        NftCollectionMetadataField,
        nullable=True,
        doc="Metadata of the collection, such as attributes, traits, etc.",
    )
    created_at = mapped_column(DateTime(timezone=True), server_default=func.now())

    telegram_chat_nft_collections = relationship(
        "TelegramChatNFTCollection",
        back_populates="nft_collection",
    )

    def __repr__(self):
        return f"<NFTCollection(address={self.address}, name={self.name})>"


class NftItem(Base):
    __tablename__ = "nft_item"

    address = mapped_column(BlockchainAddressRawField, primary_key=True)
    owner_address = mapped_column(
        ForeignKey("user_wallet.address", ondelete="CASCADE"),
        nullable=False,
    )
    collection_address = mapped_column(
        ForeignKey("nft_collection.address", ondelete="CASCADE"),
        nullable=False,
    )

    # Check NftItemMetadataDTO for expected structure
    blockchain_metadata: Mapped[BaseNftItemMetadataDTO] = mapped_column(
        NftItemMetadataField,
        nullable=True,
        doc="Metadata of the NFT, such as attributes, traits, etc.",
    )
    created_at = mapped_column(
        DateTime(timezone=True),
        server_default=func.now(),
        nullable=False,
    )
    updated_at = mapped_column(
        DateTime(timezone=True),
        server_default=func.now(),
        onupdate=func.now(),
        nullable=False,
    )

    def __repr__(self):
        return f"<NftItem(address={self.address})>"

File: lets-go/internal/models/entity/blockchain.go
package entity

import (
	"time"
)

// Jetton represents a Jetton entity in the database.
type Jetton struct {
	Address    string    `gorm:"primaryKey;column:address"`             // Blockchain Address
	Name       string    `gorm:"column:name;not null"`                 // Name of the Jetton
	Description *string   `gorm:"column:description;type:text"`         // Jetton description
	Symbol     string    `gorm:"column:symbol;not null"`               // Symbol for the Jetton
	TotalSupply int64     `gorm:"column:total_supply;not null"`         // Total supply of the Jetton
	LogoPath   string    `gorm:"column:logo_path;not null"`            // Path to the logo
	IsEnabled  bool      `gorm:"column:is_enabled;not null;default:true"` // Whether the Jetton is enabled
	CreatedAt  time.Time `gorm:"column:created_at;type:timestamp;default:current_timestamp"` // Creation time

	// One-to-Many Relationship
	TelegramChatJettons []TelegramChatJetton `gorm:"foreignKey:JettonAddress;references:Address"`
}

// TableName sets the correct table name for the Jetton entity.
func (Jetton) TableName() string {
	return "jetton"
}


// NFTCollection represents a collection of NFTs in the database.
type NFTCollection struct {
	Address          string  `gorm:"primaryKey;column:address"`            // Blockchain Address
	Name             string  `gorm:"column:name;not null"`                // Name of the Collection
	Description      *string `gorm:"column:description;type:text"`        // Description of the Collection
	LogoPath         string  `gorm:"column:logo_path;not null"`           // Logo image path
	IsEnabled        bool    `gorm:"column:is_enabled;not null;default:true"` // Whether the collection is enabled
	BlockchainMetadata []byte  `gorm:"column:blockchain_metadata;type:json"` // Serialized metadata in JSON format
	CreatedAt        time.Time `gorm:"column:created_at;type:timestamp;default:current_timestamp"` // Creation time

	// One-to-Many Relationship
	TelegramChatNftCollections []TelegramChatNFTCollection `gorm:"foreignKey:CollectionAddress;references:Address"`
}

// TableName sets the correct table name for the NFTCollection entity.
func (NFTCollection) TableName() string {
	return "nft_collection"
}


// NftItem represents an individual NFT in the database.
type NftItem struct {
	Address           string  `gorm:"primaryKey;column:address"`            // Blockchain Address
	OwnerAddress      string  `gorm:"column:owner_address;not null;index"` // Owner's blockchain address (foreign key)
	CollectionAddress string  `gorm:"column:collection_address;not null;index"` // Collection blockchain address (foreign key)
	BlockchainMetadata []byte  `gorm:"column:blockchain_metadata;type:json"` // Serialized metadata in JSON format
	CreatedAt         time.Time `gorm:"column:created_at;type:timestamp;default:current_timestamp"` // Creation time
	UpdatedAt         time.Time `gorm:"column:updated_at;type:timestamp;default:current_timestamp;onupdate:current_timestamp"`

	// Relationships
	UserWallet      UserWallet     `gorm:"foreignKey:OwnerAddress;references:Address;constraint:OnDelete:CASCADE"`
	NFTCollection   NFTCollection  `gorm:"foreignKey:CollectionAddress;references:Address;constraint:OnDelete:CASCADE"`
}

// TableName sets the correct table name for the NftItem entity.
func (NftItem) TableName() string {
	return "nft_item"
}